Mathlib/Tactic/NormNum/NatFactorial.lean

```diff
@@ -0,0 +1,120 @@
+partial def proveAscFactorial (n l : ℕ) : (result : Q(ℕ)) × Q(($n).ascFactorial $l = $result) :=
+  if l ≤ 50 then
+    let res : Q(ℕ) := mkRawNatLit (n.ascFactorial l)
+    ⟨res, (q(Eq.refl $res) : Expr)⟩
+  else
+    let m : ℕ := l / 2
+    let r : ℕ := l - m
+    let ⟨a, a_prf⟩ := proveAscFactorial n m
+    let ⟨b, b_prf⟩ := proveAscFactorial (n + m) r
+    let prf : Expr := q(asc_factorial_aux $n $m $r $a $b $a_prf $b_prf)
+    ⟨mkRawNatLit (a.natLit! * b.natLit!), prf⟩
+
+lemma isNat_factorial {n x : ℕ} (h₁ : IsNat n x) (a : ℕ) (h₂ : (0).ascFactorial x = a) :
+    IsNat (n !) a := by
+  constructor
+  simp [zero_ascFactorial, h₁.out, ← h₂]
+
+/-- Evaluates the `Nat.factorial` function. -/
+@[norm_num Nat.factorial _]
+def evalNatFactorial : NormNumExt where eval {u α} e := do
+  let .app _ (x : Q(ℕ)) ← Meta.whnfR e | failure
+  let sℕ : Q(AddMonoidWithOne ℕ) := q(instAddMonoidWithOneNat)
+  let ⟨ex, p⟩ ← deriveNat x sℕ
+  let ⟨val, ascPrf⟩ := proveAscFactorial 0 ex.natLit!
+  let ascPrf' : Q(ascFactorial 0 $ex = $val) := ascPrf
+  let prf := q(isNat_factorial $p $val $ascPrf')
+  return .isNat sℕ q($val) prf
+
+lemma isNat_ascFactorial {n x l y : ℕ} (h₁ : IsNat n x) (h₂ : IsNat l y) (a : ℕ)
+    (p : (x).ascFactorial y = a) : IsNat (n.ascFactorial l) a := by
+  constructor
+  simp [h₁.out, h₂.out, ← p]
+
+/-- Evaluates the Nat.ascFactorial function. -/
+@[norm_num Nat.ascFactorial _ _]
+def evalNatAscFactorial : NormNumExt where eval {u α} e := do
+  let .app (.app _ (x : Q(ℕ))) (y : Q(ℕ)) ← Meta.whnfR e | failure
+  let sℕ : Q(AddMonoidWithOne ℕ) := q(instAddMonoidWithOneNat)
+  let ⟨ex₁, p₁⟩ ← deriveNat x sℕ
+  let ⟨ex₂, p₂⟩ ← deriveNat y sℕ
+  let ⟨val, ascPrf⟩ := proveAscFactorial (ex₁.natLit!) (ex₂.natLit!)
+  let ascPrf' : Q(ascFactorial $ex₁ $ex₂ = $val) := ascPrf
+  let prf := q(isNat_ascFactorial $p₁ $p₂ $val $ascPrf')
+  return .isNat sℕ q($val) prf
+
+lemma isNat_descFactorial {n x l y : ℕ} (z : ℕ) (h₁ : IsNat n x) (h₂ : IsNat l y)
+    (h₃ : x = z + y) (a : ℕ) (p : z.ascFactorial y = a) : IsNat (n.descFactorial l) a := by
+  constructor
+  simp [h₁.out, h₂.out, ← p, h₃]
+  apply Nat.add_descFactorial_eq_ascFactorial
+
+lemma isNat_descFactorial_zero {n x l y : ℕ} (z : ℕ) (h₁ : IsNat n x) (h₂ : IsNat l y)
+    (h₃ : y = z + x + 1) : IsNat (n.descFactorial l) 0 := by
+  constructor
+  simp [h₁.out, h₂.out, h₃]
+
+private partial def evalNatDescFactorialNotZero {x' y' : Q(ℕ)} (x y z : Q(ℕ)) (px : Q(IsNat $x' $x))
+    (py : Q(IsNat $y' $y)) : Expr × Expr :=
+  let eq_prf : Q($x = $z + $y) := (q(Eq.refl $x) : Expr)
+  let ⟨val, ascPrf⟩ := proveAscFactorial (z.natLit!) (y.natLit!)
+  let ascPrf : Q(ascFactorial $z $y = $val) := ascPrf
+  let prf : Q(IsNat (descFactorial $x' $y') $val) :=
+    q(isNat_descFactorial $z $px $py $eq_prf $val $ascPrf)
+  (val, prf)
+
+private partial def evalNatDescFactorialZero {x' y' : Q(ℕ)} (x y z : Q(ℕ)) (px : Q(IsNat $x' $x))
+    (py : Q(IsNat $y' $y)) : Expr × Expr :=
```

Comment:
I think it would make sense to return a `Q` type here:
```suggestion
    (py : Q(IsNat $y' $y)) : (n  : Q(ℕ)) × Q(IsNat (descFactorial $x' $y') $n) :=
```